Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/EnumsAndHelperStructs.proto

```diff
@@ -0,0 +1,1226 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+syntax = "proto3";
+
+import "tools/google/protobuf/timestamp.proto";
+import "tools/google/protobuf/duration.proto";
+
+package BuildXL.Execution.Analyzer;
+
+option csharp_namespace = "BuildXL.Execution.Analyzer.Xldb";
+
+//
+// Enums
+//
+
+// Note that enum values use C++ scoping rules, meaning that enum values are siblings of their type, 
+// not children of it. Therefore, enum values must be unique within "BuildXL.Execution.Analyzer", not 
+// just within a particular enum. If there are conflicting enum values, append the name of the enum in front.
+// ie: ChooseWorkerCpu ---> PipExecutionStep_ChooseWorkerCpu
+
+enum ExecutionEventId{

```

Comment:
It really looks to me that this file ought to be auto-generated.  Otherwise, whenever someone adds any new enum constant this file will have to be manually updated.

Reply:
I agree with Aleks. That said, I would just  add a TODO here since I think that's out of scope for this PR.

---
In reply to: [313452127](https://github.com/microsoft/BuildXL/pull/724#discussion_r313452127) [](ancestors = 313452127)

Reply:
I also agree with this, the problem though is that Protobuf does not support things like inheritance or fancy maps with objects (messages in Protobuf) as keys and what not. Thus the auto-generator would have to still be maintained when things are added, but we can discuss this more. Another idea that Newton and I discussed was to instead have some kind of "validator" that could run and check if fields in the files are present here, and a user could suppress fields that should/shouldn't be in the proto file with some annotations. This way the developer making the change is more aware of what has been changed/added (and it won't mess up the numbering of items in an enum or in a message). Newton mentioned the Roslyn compiler can provide some object representation or string representation of fields so maybe we can use that. 

I'll put this as a todo for a future PR. 

---
In reply to: [314339170](https://github.com/microsoft/BuildXL/pull/724#discussion_r314339170) [](ancestors = 314339170,313452127)

Reply:
Another thing to consider is that the proto schema should be pretty static so we don't make breaking changes for things like renaming fields or refactoring. Having an automatic tool to update the schema could be counter productive to that goal in some circumstances

---
In reply to: [314401935](https://github.com/microsoft/BuildXL/pull/724#discussion_r314401935) [](ancestors = 314401935,314339170,313452127)

Reply:
Making a user story so we can discuss this at length for a future PR. 

---
In reply to: [314451213](https://github.com/microsoft/BuildXL/pull/724#discussion_r314451213) [](ancestors = 314451213,314401935,314339170,313452127)

Reply:
One thing that might make it a bit more maintainable long term would be to split each enum/message in its own file.
For future archeolygy if we have to see why feelds were added / removed it will be easier to look in the history.

---
In reply to: [314551183](https://github.com/microsoft/BuildXL/pull/724#discussion_r314551183) [](ancestors = 314551183,314451213,314401935,314339170,313452127)

Reply:
This is an interesting idea, but then there will be a bunch of import statements whenever we need to import 1 or more enums. The question then stands, would we also want to split each message into its own file, and would that lead to too much clutter with a user having to [possibly] maintain a bunch of files rather than a small handful?

---
In reply to: [314809335](https://github.com/microsoft/BuildXL/pull/724#discussion_r314809335) [](ancestors = 314809335,314551183,314451213,314401935,314339170,313452127)